compiler/rustc_hir/src/hir.rs

```diff
@@ -1811,6 +1811,41 @@ impl Expr<'_> {
         }
     }
 
+    /// Whether this and the `other` expression are the same for purposes of an indexing operation.
+    ///
+    /// This is only used for diagnostics to see if we have things like `foo[i]` where `foo` is
+    /// borrowed multiple times with `i`.
+    pub fn equals(&self, other: &Expr<'_>) -> bool {
```

Comment:
The naming seems confusing as it doesn't specify that this is only in terms of indexing operations. Perhaps rename to `equivalent_for_indexing`?